From c8bbd5a9776fa03811ef11f73aa598059f64e002 Mon Sep 17 00:00:00 2001
From: Stuart Prescott <stuart@debian.org>
Date: Wed, 4 Jun 2014 00:49:01 +1000
Subject: [PATCH] Allow whitespace-only lines to separate paragraphs

iter_paragraphs should split paragraphs on whitespace-only lines.

Paragraphs should be separated by a blank line but policy permits parsers to
permit whitespace only lines too.

Closes #715558.
---
 lib/debian/deb822.py |  4 +++-
 tests/test_deb822.py | 24 ++++++++++++++++++++++++
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/lib/debian/deb822.py b/lib/debian/deb822.py
index 98bf9f3..ed933e4 100644
--- a/lib/debian/deb822.py
+++ b/lib/debian/deb822.py
@@ -627,7 +627,9 @@ class Deb822(Deb822Dict):
         gpg_post_lines = []
         state = b'SAFE'
         gpgre = re.compile(br'^-----(?P<action>BEGIN|END) PGP (?P<what>[^-]+)-----$')
-        blank_line = re.compile(b'^$')
+        # Include whitespace-only lines in blank lines to split paragraphs.
+        # (see #715558)
+        blank_line = re.compile(b'^\s*$')
         first_line = True
 
         for line in sequence:
diff --git a/tests/test_deb822.py b/tests/test_deb822.py
index a989a32..614bd5a 100755
--- a/tests/test_deb822.py
+++ b/tests/test_deb822.py
@@ -445,6 +445,30 @@ class TestDeb822(unittest.TestCase):
                 self.assertWellParsed(d, PARSED_PACKAGE)
             self.assertEqual(count, 2)
 
+    def test_iter_paragraphs_with_extra_whitespace(self):
+        """ Paragraphs not elided when stray whitespace is between
+
+        From policy ยง5.1:
+
+            The paragraphs are separated by empty lines. Parsers may accept
+            lines consisting solely of spaces and tabs as paragraph separators,
+            but control files should use empty lines.
+
+        On the principle of "be strict in what you send; be generous in
+        what you receive", deb822 should permit such extra whitespace between
+        deb822 stanzas.
+
+        See #715558 for further details.
+        """
+        for extra_space in (" ", "  ", "\t"):
+            text = (UNPARSED_PACKAGE + '%s\n' % extra_space
+                        + UNPARSED_PACKAGE).splitlines()
+            count = 0
+            for d in deb822.Deb822.iter_paragraphs(text):
+                count += 1
+            self.assertEqual(count, 2,
+                        "Wrong number paragraphs were found: %d != 2" % count)
+
     def _test_iter_paragraphs(self, filename, cls, **kwargs):
         """Ensure iter_paragraphs consistency"""
         
-- 
1.9.1

